lib/model/message.dart

```diff
@@ -1,21 +1,332 @@
+    required super.localMessageId,
+    required super.selfUserId,
+    required super.timestamp,
+    required this.conversation,
+    required super.content,
+  });
+
+  @override
+  final StreamConversation conversation;
+}
+
+class DmOutboxMessage extends OutboxMessage<DmConversation> {
+  DmOutboxMessage({
+    required super.localMessageId,
+    required super.selfUserId,
+    required super.timestamp,
+    required this.conversation,
+    required super.content,
+  }) : assert(conversation.allRecipientIds.contains(selfUserId));
+
+  @override
+  final DmConversation conversation;
+}
+
+/// Manages the outbox messages portion of [MessageStore].
+mixin _OutboxMessageStore on PerAccountStoreBase {
+  late final UnmodifiableMapView<int, OutboxMessage> outboxMessages =
+    UnmodifiableMapView(_outboxMessages);
+  final Map<int, OutboxMessage> _outboxMessages = {};
+
+  /// A map of timers to show outbox messages after a delay,
+  /// indexed by [OutboxMessage.localMessageId].
+  ///
+  /// If the send message request fails within the time limit,
+  /// the outbox message's timer gets removed and cancelled.
+  final Map<int, Timer> _outboxMessageDebounceTimers = {};
+
+  /// A map of timers to update outbox messages state to
+  /// [OutboxMessageState.waitPeriodExpired] after a delay,
+  /// indexed by [OutboxMessage.localMessageId].
+  ///
+  /// If the send message request fails within the time limit,
+  /// the outbox message's timer gets removed and cancelled.
+  final Map<int, Timer> _outboxMessageWaitPeriodTimers = {};
+
+  /// A fresh ID to use for [OutboxMessage.localMessageId],
+  /// unique within this instance.
+  int _nextLocalMessageId = 0;
+
+  /// As in [MessageStoreImpl._messageListViews].
+  Set<MessageListView> get _messageListViews;
+
+  /// As in [MessageStoreImpl._disposed].
+  bool get _disposed;
+
+  /// Update the state of the [OutboxMessage] with the given [localMessageId],
+  /// and notify listeners if necessary.
+  ///
+  /// The outbox message with [localMessageId] must exist.
+  void _updateOutboxMessage(int localMessageId, {
+    required OutboxMessageState newState,
+  }) {
+    assert(!_disposed);
+    final outboxMessage = outboxMessages[localMessageId];
+    if (outboxMessage == null) {
+      throw StateError(
+        'Removing unknown outbox message with localMessageId: $localMessageId');
+    }
+    final oldState = outboxMessage.state;
+    // See [OutboxMessageState] for valid state transitions.
+    final isStateTransitionValid = newState != oldState && switch (newState) {
+      OutboxMessageState.hidden => false,
+      OutboxMessageState.waiting =>
+        oldState == OutboxMessageState.hidden,
+      OutboxMessageState.waitPeriodExpired =>
+        oldState == OutboxMessageState.waiting,
+      OutboxMessageState.failed =>
+        oldState == OutboxMessageState.hidden
+        || oldState == OutboxMessageState.waiting
+        || oldState == OutboxMessageState.waitPeriodExpired,
```

Comment:
nit: equivalently,
```suggestion
      OutboxMessageState.failed => true,
```

right? Given the `newState != oldState` above.

Alternatively, I think with this condition in place here, that inequality check is redundant.